### PR TITLE
Leader election rate

### PR DESCRIPTION
**Pull request by sudiksha27**
**Goals (and why)**:
Add a health check on leader election rate. This will be useful for moving TimeLock to blue-green upgrade model.

**Implementation Description (bullets)**:

- Move LeadershipEvents metrics to metric-schema.

- Add a health check on rate of leader health checks.

**Testing (What was existing testing like?  What have you done to improve it?)**:

- Added a test for leader election health check

**Concerns (what feedback would you like?)**:

- Is the HealthCheck threshold estimate reasonable? 

**Where should we start reviewing?**:

- LeaderElectionHealthCheck, LeadershipEvents

**Priority (whenever / two weeks / yesterday)**:

- Before end of week

>

### CODE REVIEW

File: leader-election-impl/src/main/java/com/palantir/leader/health/LeaderElectionHealthCheck.java [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+import com.palantir.leader.LeaderElectionServiceMetrics;
+import com.palantir.tritium.metrics.registry.TaggedMetricRegistry;
+
+public class LeaderElectionHealthCheck {
+    private static final double MAX_ALLOWED_5_MIN_RATE = 5;
+    private static final double MAX_ALLOWED_15_MIN_RATE = 15;
+
+    private final LeaderElectionServiceMetrics leaderElectionServiceMetrics;
+
+    public LeaderElectionHealthCheck(TaggedMetricRegistry taggedMetricRegistry) {
```
Comment (jkozlowski) [inline leader-election-impl/src/main/java/com/palantir/leader/health/LeaderElectionHealthCheck.java:28]: How hard would it be to add a test?

File: leader-election-impl/src/main/java/com/palantir/leader/health/LeaderElectionHealthCheck.java [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+ * limitations under the License.
+ */
+
+package com.palantir.leader.health;
+
+import com.palantir.leader.LeaderElectionServiceMetrics;
+import com.palantir.tritium.metrics.registry.TaggedMetricRegistry;
+
+public class LeaderElectionHealthCheck {
+    private static final double MAX_ALLOWED_5_MIN_RATE = 5;
```
Comment (jkozlowski) [inline leader-election-impl/src/main/java/com/palantir/leader/health/LeaderElectionHealthCheck.java:23]: This means 5 in the last minute? or 5 in the last 5 minutes?

File: leader-election-impl/src/main/java/com/palantir/leader/health/LeaderElectionHealthCheck.java (hunk 1 above)
Comment (sudiksha27) [reply leader-election-impl/src/main/java/com/palantir/leader/health/LeaderElectionHealthCheck.java]: Test added for the health check.

File: leader-election-impl/src/main/java/com/palantir/leader/health/LeaderElectionHealthCheck.java (hunk 2 above)
Comment (sudiksha27) [reply leader-election-impl/src/main/java/com/palantir/leader/health/LeaderElectionHealthCheck.java]: This constant is incorrect; i am fixing this.

File: leader-election-impl/src/main/java/com/palantir/leader/health/LeaderElectionHealthCheck.java [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.leader.health;
+
+import com.palantir.leader.LeaderElectionServiceMetrics;
+
+public class LeaderElectionHealthCheck {
+    private static final double MAX_ALLOWED_LAST_5_MINUTE_RATE = 1;
```
Comment (jeremyk-91) [inline leader-election-impl/src/main/java/com/palantir/leader/health/LeaderElectionHealthCheck.java:22]: This feels a bit high: one election per second is pretty unstable. Based on metrics we've seen internally I think we want this to be maybe one in 30 seconds?

File: leader-election-impl/src/main/java/com/palantir/leader/LeadershipEvents.java [hunk 4]
```diff
@@ -35,64 +34,49 @@
...
-    private final Meter gainedLeadership;
-    private final Meter lostLeadership;
-    private final Meter noQuorum;
-    private final Meter proposedLeadership;
-    private final Meter proposalFailure;
-    private final Meter leaderPingFailure;
-    private final Meter leaderPingTimeout;
-    private final Meter leaderPingReturnedFalse;
     private final Object[] contextArgs;
+    private final LeaderElectionServiceMetrics leaderElectionServiceMetrics;
```
Comment (jeremyk-91) [inline leader-election-impl/src/main/java/com/palantir/leader/LeadershipEvents.java:38]: Nice work!

File: leader-election-impl/src/test/java/com/palantir/leader/LeadershipElectionCheckTest.java [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+
+import com.codahale.metrics.Clock;
+import com.codahale.metrics.Meter;
+import com.palantir.leader.health.LeaderElectionHealthCheck;
+import com.palantir.leader.health.LeaderElectionHealthStatus;
+import com.palantir.tritium.metrics.registry.TaggedMetricRegistry;
+
+public class LeadershipElectionCheckTest {
+    private final FakeTimeClock fakeTimeClock = new FakeTimeClock();
+    private final TaggedMetricRegistry registry1 = mock(TaggedMetricRegistry.class);
```
Comment (jeremyk-91) [inline leader-election-impl/src/test/java/com/palantir/leader/LeadershipElectionCheckTest.java:39]: nit: `registry`

File: leader-election-impl/src/test/java/com/palantir/leader/LeadershipElectionCheckTest.java [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+    @Test
+    public void shouldBeUnhealthyForMoreThanOneLeaderElectionPerMinute() {
+        markLeaderElectionsAtSpecifiedInterval(7, 30);
+        assertThat(leaderElectionHealthCheck.leaderElectionRateHealthStatus())
+                .isEqualTo(LeaderElectionHealthStatus.UNHEALTHY);
+    }
+
+    public void markLeaderElectionsAtSpecifiedInterval(int leaderElectionCount, long timeIntervalInSeconds) {
+        // First tick
+        fakeTimeClock.advance(6, TimeUnit.SECONDS);
```
Comment (jeremyk-91) [inline leader-election-impl/src/test/java/com/palantir/leader/LeadershipElectionCheckTest.java:67]: Sorry - despite the comment I'm not sure what the significance of the 6 seconds here is?

File: leader-election-impl/src/test/java/com/palantir/leader/LeadershipElectionCheckTest.java [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+
+        IntStream.range(0, leaderElectionCount).forEach(idx -> {
```
Comment (jeremyk-91) [inline leader-election-impl/src/test/java/com/palantir/leader/LeadershipElectionCheckTest.java:69]: ❤️

File: leader-election-impl/src/test/java/com/palantir/leader/LeadershipElectionCheckTest.java [hunk 8]
```diff
@@ -0,0 +1,89 @@
...
+            new LeaderElectionHealthCheck(leaderElectionServiceMetrics);
+
+    @Before
+    public void setup() {
+        when(registry1.meter(any())).thenReturn(new Meter(fakeTimeClock));
+    }
+
+    @Test
+    public void shouldBeHealthyForOneLeaderElectionPerMinute() {
+        markLeaderElectionsAtSpecifiedInterval(3, 60);
```
Comment (jeremyk-91) [inline leader-election-impl/src/test/java/com/palantir/leader/LeadershipElectionCheckTest.java:52]: nit: I'd pass Duration into markLeaderElections(...) for readability

File: leader-election-impl/src/test/java/com/palantir/leader/LeadershipElectionCheckTest.java [hunk 9]
```diff
@@ -0,0 +1,89 @@
...
+    }
+
+    @Test
+    public void shouldBeUnhealthyForMoreThanOneLeaderElectionPerMinute() {
+        markLeaderElectionsAtSpecifiedInterval(7, 30);
+        assertThat(leaderElectionHealthCheck.leaderElectionRateHealthStatus())
+                .isEqualTo(LeaderElectionHealthStatus.UNHEALTHY);
+    }
+
+    public void markLeaderElectionsAtSpecifiedInterval(int leaderElectionCount, long timeIntervalInSeconds) {
```
Comment (jeremyk-91) [inline leader-election-impl/src/test/java/com/palantir/leader/LeadershipElectionCheckTest.java:65]: nit: private?

File: leader-election-impl/src/main/java/com/palantir/leader/health/LeaderElectionHealthCheck.java [hunk 10]
```diff
@@ -0,0 +1,34 @@
...
+    private static final double MAX_ALLOWED_LAST_5_MINUTE_RATE = 0.016;
+
+    private final LeaderElectionServiceMetrics leaderElectionServiceMetrics;
+
+    public LeaderElectionHealthCheck(LeaderElectionServiceMetrics leaderElectionServiceMetrics) {
+        this.leaderElectionServiceMetrics = leaderElectionServiceMetrics;
+    }
+
+    public LeaderElectionHealthStatus leaderElectionRateHealthStatus() {
+        return leaderElectionServiceMetrics.proposedLeadership().getFiveMinuteRate() < MAX_ALLOWED_LAST_5_MINUTE_RATE
```
Comment (jeremyk-91) [inline leader-election-impl/src/main/java/com/palantir/leader/health/LeaderElectionHealthCheck.java:31]: nit: the name of the constant implies `<=`?

File: leader-election-impl/src/main/metrics/leader-election.yml [hunk 11]
```diff
@@ -0,0 +1,31 @@
+options:
+  javaPackage: 'com.palantir.leader'
+
+namespaces:
+  leaderElectionService:
+    docs: Metrics helpful for TimeLock adjudication.
```
[the remark this answers is not in the document]
Comment (jeremyk-91) [reply leader-election-impl/src/main/metrics/leader-election.yml:6]: (I think we still need to fix this in round 2)

File: leader-election-impl/src/main/metrics/leader-election.yml [hunk 12]
```diff
@@ -0,0 +1,31 @@
...
+        docs: Number of times(per second) leadership was lost.
+      noQuorum:
+        type: meter
+        docs: Number of times(per second) there was no quorum while determining leadership status.
+      proposedLeadership:
+        type: meter
+        docs: Number of times(per second) leadership was proposed.
+      proposalFailure:
+        type: meter
+        docs: Number of times(per second) leadership could not be gained after proposal.
```
Comment (jeremyk-91) [inline leader-election-impl/src/main/metrics/leader-election.yml:22]: We should standardise between `Number of times(per second)` and `Rate`. I think I slightly prefer the former - or maybe `Rate (frequency per second)`?

**Pull request merged**